packages/daemon/src/daemon-node-powers.js

```diff
@@ -279,111 +375,208 @@ export const makePowers = ({
-   */
-  const makeFileReader = path => {
-    const nodeReadStream = fs.createReadStream(path);
-    return makeNodeReader(nodeReadStream);
+  const makeHashedContentReadeableBlob = (statePath, sha512) => {
+    const storageDirectoryPath = diskPowers.joinPath(
+      statePath,
+      'store-sha512',
+    );
+    const storagePath = diskPowers.joinPath(storageDirectoryPath, sha512);
+    const stream = async () => {
+      const reader = diskPowers.makeFileReader(storagePath);
+      return makeReaderRef(reader);
+    };
+    const text = async () => {
+      return diskPowers.readFileText(storagePath);
+    };
+    const json = async () => {
+      return JSON.parse(await text());
+    };
+    return { stream, text, json }
   };
 
-  /**
-   * @param {string} path
-   */
-  const makeFileWriter = path => {
-    const nodeWriteStream = fs.createWriteStream(path);
-    return makeNodeWriter(nodeWriteStream);
-  };
+  const makeHashedContentWriter = async (statePath) => {
+    const storageDirectoryPath = diskPowers.joinPath(
+      statePath,
+      'store-sha512',
+    );
+    await diskPowers.makePath(storageDirectoryPath);
+
+    // Pump the reader into a temporary file and hash.
+    // We use a temporary file to avoid leaving a partially writen object,
+    // but also because we won't know the name we will use until we've
+    // completed the hash.
+    const digester = makeSha512();
+    const storageId512 = await randomHex512();
+    const temporaryStoragePath = diskPowers.joinPath(
+      storageDirectoryPath,
+      storageId512,
+    );
+    const fileWriter = diskPowers.makeFileWriter(temporaryStoragePath);
+    const sha512Kit = makePromiseKit();
+
+    /** @type {import('@endo/stream').Writer<Uint8Array>} */
+    const hashedFileWriter = {
+      async next(chunk) {
+        digester.update(chunk);
+        return fileWriter.next(chunk);
+      },
+      async return() {
+        // Finish writing the file.
+        const result = await fileWriter.return(undefined);
+        // Calculate hash.
+        const sha512 = digester.digestHex();
+        // Finish with an atomic rename.
+        const storagePath = diskPowers.joinPath(storageDirectoryPath, sha512);
+        await diskPowers.renamePath(temporaryStoragePath, storagePath);
+        // Notify the caller of the hash.
+        sha512Kit.resolve(sha512);
+        return result;
+      },
+      async throw(error) {
+        return fileWriter.throw(error);
+      },
+      [Symbol.asyncIterator]() {
+        return this;
+      },
+    };
 
-  /**
-   * @param {string} path
-   * @param {string} text
-   */
-  const writeFileText = async (path, text) => {
-    await fs.promises.writeFile(path, text);
-  };
+    const getSha512Hex = () => {
+      return sha512Kit.promise;
+    }
 
-  /**
-   * @param {string} path
-   */
-  const readFileText = async path => {
-    return fs.promises.readFile(path, 'utf-8');
+    return { writer: hashedFileWriter, getSha512Hex };
```

Comment:
better interface for getting the hash on completion?

Reply:
Yeah, I’m feeling better about abstracting this as “makeContentSha512Store” with `store(readable) => Promise<hash: string>` and `fetch(hash: string)`